### DISCUSSION TITLE
Thermostat device matter

### DISCUSSION THREAD

**Discussion by goeran03** [arendst/Tasmota#20373] [Ideas]
Is it planned to implement a thermostat device in matter? A virtual device at least would be probably give the possibility to be flexible. A simple behavior with a thermo sensor, a relay and a given switching point could be realized that way with scripts I guess, and more complex solutions as well. If there is thermostat functionality already in Tasmota , this could be linked to matter.

**Comment by sfromis** [comment]
Tasmota does indeed have a [Thermostat](https://tasmota.github.io/docs/Thermostat/) feature. As I'm not using this, IDK how simple it would be to integrate.

I suppose that it could be feasible by mapping Matter inputs/actions to the corresponding "standard Tasmota" ways of handling these, and then letting the Thermostat feature operate on that. Just uninformed speculation of course :smile: 

It would be another case with a Matter-compatible thermostat device.

**Comment by s-hadinger** [comment]
@goeran03  Are you already using the Thermostat feature?

**Reply by goeran03** [reply]
Yes, using a tasmota esp32 in a very simple way using just thermo sensor and relay function of it. But the actual algorithm is done on another device using nodered to have HomeKit connection possible. 
It would be great if I could integrate everything including matter interface to the tasmota device.
Thanks for looking into it.

**Reply by sfromis** [reply]
The question was about whether you are using the specific Thermostat feature of Tasmota (which I posted a link to), instead of just "a very simple way using just thermo sensor and relay function of it".

The Thermostat feature provides a lot of commands to provide better results than simply reacting to current temperature.

**Reply by goeran03** [reply]
Not yet

**Reply by benjaminnestler** [reply]
I'm currently working on a thermostat cluster integration but in the first step not for the tasmota Thermostat feature. I want to adopt a Tuya based thermostat to the thermostat cluster of Matter.

**Reply by s-hadinger** [reply]
Awesome

**Reply by benjaminnestler** [reply]
Some pictures to show my progress 

|<img src="https://github.com/arendst/Tasmota/assets/101095581/c4883ba0-a4dd-4687-84fc-6445a6c82b90" alt="image" width="300" height="auto"> | <img src="https://github.com/arendst/Tasmota/assets/101095581/c1a1ab82-116f-4ae0-aaed-e49d5c5a7cd0" alt="image" width="300" height="auto"> |
| <img src="https://github.com/arendst/Tasmota/assets/101095581/cfcd0091-66fc-44f4-9f29-e6495dbd232e" alt="image" width="300" height="auto"> | <img src="https://github.com/arendst/Tasmota/assets/101095581/bd7b53bf-4391-4de6-b1c7-d41840649ee2" alt="image" width="300" height="auto"> |
| <img src="https://github.com/arendst/Tasmota/assets/101095581/8dc0ec6e-b459-40b3-b841-67b434c76d92" alt="image" width="300" height="auto"> | <img src="https://github.com/arendst/Tasmota/assets/101095581/1fd28436-8be9-4a8c-be0c-d3bc09057e2f" alt="image" width="300" height="auto"> |
| <img src="https://github.com/arendst/Tasmota/assets/101095581/a273a5f1-c9da-4055-9dd5-a7f5afceea02" alt="image" width="300" height="auto"> | <img src="https://github.com/arendst/Tasmota/assets/101095581/d5b7843e-10a7-4eb8-aa26-8367af8ad4a0" alt="image" width="300" height="auto"> |
|---|---|

**Reply by vincedarley** [reply]
This looks very interesting. I'm currently in the situation of using dumb relays (Sonoff 4CH-R3) as thermostats with Home Assistant (HASS)'s Basic Thermostat doing all of the control logic (given a Xiaomi temperature sensor reading in each area).

I'm looking at how best to make this more robust so that even if HASS is down, the system generally operates very comfortably.  As I see it, most more sophisticated thermostats have 4 different modes (variously named Comfort, Eco/Night, Away, Vacation/Frost), each of which has its own set-point. And then there is a "heating scheme" which is a schedule (perhaps 24 hour, perhaps 1 week long), which says what mode to use on the given day/time. Once the mode is set, the temperature set-point is known.

There seem to be 4 options of increasing sophistication:
1. Tasmota only knows about a temperature set-point, which HASS tells it (~Hourly)
2. Tasmota knows about a temperature set-point per mode, and HASS tells it the mode (~Several times per day)
3. Tasmota knows about a 24-hour “heating scheme” and HASS tells it the scheme to use (~2 times per week?).
4. Tasmota knows about a full week’s “heating scheme”, including weekend or day of week variations, and HASS only tells it to change for special purposes (e.g. homeowner is on vacation for a week, so HASS only needed very rarely).

As we go down this list we move from 1,2 where HASS is a critical part of the system (= basic daily heating functionality will be uncomfortable if HASS goes down at the wrong time) to 3 and especially 4 where HASS is non-critical, and the user’s comfort would not be impacted if HASS goes down.  In 4, and perhaps 3, you could imagine a user being willing to open the web portal to the Tasmota device (or HomeKit/Matter on their phone) to make occasional adjustments and HASS not even being required at all.

Of course it would be nice to support temporary modifications to 3&4 where a HASS automation or a user could tell the Tasmota thermostat - “Boost the temperature in this room +1 degree until the next scheduled change" or “We’re not going to be home today, the house can be in Eco mode for the next 6 hours”.  On expiry the thermostat would revert to its standard programming. Some of the KNX thermostats I've been reading about support this kind of thing.

Any clarity on how far in the above I can get with Tasmota today?

**Comment by Cptmeme** [comment]
Are there any updates regarding this? I am interested in integrating a tuya space heater I have into matter.

**Comment by Fabrizz** [comment]
@benjaminnestler that implementation looks great. Im really interested in having climate support, is there any updates on that?

Many thanks

**Reply by sfromis** [reply]
While I can't speak for him, I can observe that the first screenshot looks a lot like this project:
https://github.com/benjaminnestler/berryTasmotaTuyaDecode
